pennylane/debugging/operation_checker.py

```diff
@@ -0,0 +1,833 @@
+        if matrix is None or eigvals is None:
+            return
+        mat_eigvals = np.linalg.eigvals(matrix)
+        if not np.allclose(mat_eigvals, eigvals):
+            self.print_(
+                f"The eigenvalues of the matrix and the stored eigvals for {self.tmp['name']} do not match.",
+                "error",
+            )
+        return
+
+    def _check_diag_gates(self, diag_gates, matrix, eigvals, op):
+        """Check that the diagonalizing gates attributed to an operation
+        produce a diagonal matrix, and that it has the correct eigenvalues."""
+        if diag_gates is None or matrix is None:
+            return
+        if diag_gates == []:
+            diag_mat = np.eye(matrix.shape[0])
+        else:
+            with qml.tape.QuantumTape() as tape:
+                [op.queue() for op in diag_gates]
+            diag_mat = qml.matrix(tape)
+
+        diagonalized = diag_mat @ matrix @ diag_mat.conj().T
+        if not is_diagonal(diagonalized, atol=self.tol):
+            self.print_(
+                f"The diagonalizing gates do not diagonalize the matrix for {self.tmp['name']}.",
+                "error",
+            )
+            return
+        if eigvals is not None and not np.allclose(
+            np.sort(eigvals), np.sort(np.diag(diagonalized))
+        ):
+            self.print_(
+                "The diagonalizing gates diagonalize the matrix but produce wrong "
+                f"eigenvalues for {self.tmp['name']}.",
+                "error",
+            )
+        return
+
+    def _check_basis(self, matrix, instance):
+        """Check that a matrix attributed to an operation is diagonal in the basis
+        indicated by that operation's ``basis`` property."""
+        try:
+            basis = instance.basis
+        except AttributeError:
+            basis = None
+
+        if basis is None or matrix is None:
+            return
+
+        if basis == "X":
+            diag_gates = [qml.Hadamard]
+        elif basis == "Y":
+            diag_gates = [qml.PauliZ, qml.S, qml.Hadamard]
+        elif basis == "Z":
+            diag_gates = [qml.Identity]
+
+        target_wires = qml.wires.Wires.unique_wires([instance.wires, instance.control_wires])
+        with qml.tape.QuantumTape() as tape:
+            for w in target_wires:
+                [diag_gate(wires=w) for diag_gate in diag_gates]
+
+        diag_mat = qml.operation.expand_matrix(qml.matrix(tape), target_wires, instance.wires)
+        if not is_diagonal(diag_mat @ matrix @ diag_mat.conj().T, atol=self.tol):
+            self.print_(
+                f"The operation {self.tmp['name']} is not diagonal in the provided basis",
+                "error",
+            )
+
+    def _check_differentiability(self, op, parameters, wires):
+        """Check that an operation is differentiable if it is marked to be, and that
+        the correct derivative is produced. Note that this can be computationally expensive."""
```

Comment:
There is another approach we could take here, which is to check that `qml.matrix()` (or even other representations!) are differentiable. This is required to support backprop, since `default.qubit` simply does `qml.matrix(op) @ state` in backprop mode, and assumes that the matrix was differentiable.

You could even extend the `qml.matrix()` test to check for differentiability in multiple interfaces

Reply:
I implemented the qml.matrix-based check for all interfaces.